cli/src/semgrep/output.py

```diff
@@ -111,11 +112,37 @@ def _build_time_json(
     )
 
 
+# This class is the internal representation of OutputSettings below.
+# Since it is internal it can change as much as necisarry to make
+# typchecking more accurate and enforce invariants.
+class NormalizedOutputSettings(NamedTuple):
+    # Immutable List of OutputDestination x OutputFormat
+    outputs: Tuple[Tuple[Optional[str], OutputFormat], ...]
+    output_per_finding_max_lines_limit: Optional[int]
+    output_per_line_max_chars_limit: Optional[int]
+    error_on_findings: bool
+    verbose_errors: bool
+    strict: bool
+    output_time: bool
+    timeout_threshold: int
+    dataflow_traces: bool
+
+    def has_output_format(self, other: OutputFormat) -> bool:
+        return bool(sum(1 for (_, fmt) in self.outputs if other == fmt))
+
+    def has_text_output(self) -> bool:
+        return self.has_output_format(OutputFormat.TEXT)
+
+
 # WARNING: this class is unofficially part of our external API. It can be passed
 # as an argument to our official API: 'semgrep_main.invoke_semgrep'. Try to minimize
 # changes to this API, and make them backwards compatible, if possible.
 class OutputSettings(NamedTuple):
-    output_format: OutputFormat
+    # Immutable List of OutputDestination x OutputFormat.
+    # It is optional to maintain backwards compatibility.
```

Comment:
Personally I really wish didn't need to be backward compatible, so you can just use `outputs` and not need to have `output_format` floating around.

The needed for backward compatibility seems to be a requirement from 3 years ago: https://github.com/semgrep/semgrep/commit/3737d0faf5028cf9bcfe7878797725b68db6af11

Can anyone (cc reviewers of that PR @aryx @emjin @brendongo ) give us more context about `semgrep_main.invoke_semgrep`? Can we safely break compatibility here, or no?

Reply:
Even with the reply on [this thread](https://semgrepinc.slack.com/archives/C05STGMM53Q/p1713312640336109) that we can break this interface I would still like to land the code that I have working now, and clean it up in a later PR. This task is a contractual commit for the end of April which means that it needs to get into next week's release.

Reply:
I agree that we should get this PR in first.